### PR TITLE
Update types for Mithril 2.0

### PR DESCRIPTION
**Pull request by spacejack**
Mithril 2.0 has been released and these are the updated type definitions.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/MithrilJS/mithril.js/releases/tag/v2.0.1
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Comment by spacejack** [comment]
It looks like the checks are failing because this is breaking mithril-global (for Mithril 1.x.)
I haven't yet figured out how to update mithril-global types to work with the new Stream single export type, but I'm hoping this specific issue shouldn't hold back these types for 2.0 from being published.

**Comment by spacejack** [comment]
I have a working solution for [mithril-global](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/mithril-global) 2.0 types.
Should I commit that change to this PR or create a separate branch/PR?

**Comment by isiahmeadows** [comment]
@spacejack It might be easier/better to include that in this PR so you aren't confusing their bots as much.

File: types/mithril/index.d.ts [hunk 1]
```diff
@@ -136,11 +135,13 @@ declare namespace Mithril {
...
 		/** If false, redraws mounted components upon completion of the request. If true, it does not. */
 		background?: boolean;
+		/** Milliseconds a request can take before automatically being terminated. */
+		timeout?: number;
 	}
 
 	interface JsonpOptions {
 		/** The data to be interpolated into the URL and serialized into the querystring. */
-		data?: any;
+		params?: { [id: string]: any };
```
Comment (isiahmeadows) [inline types/mithril/index.d.ts]: Don't forget `body:` (unless of course you have it specified elsewhere).

**Review by dead-claudia — CHANGES_REQUESTED** [review]

File: types/mithril/index.d.ts (hunk 1 above)
Comment (spacejack) [reply types/mithril/index.d.ts]: Okay, good catch.

File: types/mithril/stream/index.d.ts [hunk 2]
```diff
@@ -1,40 +1,49 @@
...
+	/** Creates a stream whose value is the array of values from an array of streams. */
+	export function merge(streams: Array<Stream<any>>): Stream<any[]>;
+	/** Creates a new stream with the results of calling the function on every incoming stream with and accumulator and the incoming value. */
+	export function scan<T, U>(fn: (acc: U, value: T) => U, acc: U, stream: Stream<T>): Stream<U>;
+	/** Takes an array of pairs of streams and scan functions and merges all those streams using the given functions into a single stream. */
+	export function scanMerge<T, U>(pairs: Array<[Stream<T>, (acc: U, value: T) => U]>, acc: U): Stream<U>;
+	/** Takes an array of pairs of streams and scan functions and merges all those streams using the given functions into a single stream. */
+	export function scanMerge<U>(pairs: Array<[Stream<any>, (acc: U, value: any) => U]>, acc: U): Stream<U>;
+	/** A special value that can be returned to stream callbacks to skip execution of downstreams. */
+	export const SKIP: any;
```
Comment (isiahmeadows) [inline types/mithril/stream/index.d.ts]: Let's lie and make this `unique symbol` for now, [until TS adds `unique` types down the road that aren't just `unique symbol`s](https://github.com/microsoft/TypeScript/issues/202#issuecomment-437444382).

```suggestion
	export const SKIP: unique symbol;
```

Type safety > correctness, especially for something people shouldn't be `typeof` checking normally. It also lets us change the callback's return type of `map` to `T | SKIP` without turning the whole thing into an `any`.

File: types/mithril/stream/index.d.ts [hunk 3]
```diff
@@ -1,40 +1,49 @@
...
 }
 
-declare const Stream: Stream.Static;
```
Comment (isiahmeadows) [inline types/mithril/stream/index.d.ts]: [I like it](https://gph.is/2ar5hhh)

**Review by dead-claudia — APPROVED** [review]

**Review by dead-claudia — CHANGES_REQUESTED** [review]
(Wrong button - see previous review)

File: types/mithril/stream/index.d.ts [hunk 4]
```diff
@@ -1,40 +1,49 @@
...
-		/** Takes an array of pairs of streams and scan functions and merges all those streams using the given functions into a single stream. */
-		scanMerge<T, U>(pairs: Array<[Stream<T>, (acc: U, value: T) => U]>, acc: U): Stream<U>;
-		/** Takes an array of pairs of streams and scan functions and merges all those streams using the given functions into a single stream. */
-		scanMerge<U>(pairs: Array<[Stream<any>, (acc: U, value: any) => U]>, acc: U): Stream<U>;
-		/** A special value that can be returned to stream callbacks to halt execution of downstreams. */
-		readonly HALT: any;
-	}
+declare namespace Stream {
+	/** Creates a computed stream that reactively updates if any of its upstreams are updated. */
+	export function combine<T>(combiner: (...streams: any[]) => T, streams: Array<Stream<any>>): Stream<T>;
```
[the remark this answers is not in the document]
Comment (spacejack) [reply types/mithril/stream/index.d.ts]: Okay. We'll need to bump the compiler version for this. But I guess that's fine being Mithril V2.

File: types/mithril/index.d.ts [hunk 5]
```diff
@@ -136,11 +135,15 @@ declare namespace Mithril {
...
+		/** The data to be serialized into the request body. */
+		body?: any;
```
[the remark this answers is not in the document]
Comment (spacejack) [reply types/mithril/index.d.ts]: This now prevents number values for body, but I suppose those were just auto-coerced to strings.

File: types/mithril/stream/index.d.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (spacejack) [reply types/mithril/stream/index.d.ts]: I think I'll need some help with this one. With your signature, the combiner callback `streams` arguments types are unwrapped values, but they should be streams.

File: types/mithril/stream/index.d.ts [hunk 6]
```diff
@@ -1,40 +1,49 @@
...
+	/** Combines the values of one or more streams into a single stream that is updated whenever one or more of the sources are updated */
+	export function lift<A, Z>(fn: (a: A) => Z, s: Stream<A>): Stream<Z>;
+	export function lift<A, B, Z>(fn: (a: A, b: B) => Z, sa: Stream<A>, sb: Stream<B>): Stream<Z>;
+	export function lift<A, B, C, Z>(fn: (a: A, b: B, c: C) => Z, sa: Stream<A>, sb: Stream<B>, sc: Stream<C>): Stream<Z>;
+	export function lift<A, B, C, D, Z>(fn: (a: A, b: B, c: C, d: D) => Z, sa: Stream<A>, sb: Stream<B>, sc: Stream<C>, sd: Stream<D>): Stream<Z>;
+	export function lift<A, B, C, D, E, Z>(fn: (a: A, b: B, c: C, d: D, e: E) => Z, sa: Stream<A>, sb: Stream<B>, sc: Stream<C>, sd: Stream<D>, se: Stream<E>): Stream<Z>;
+	export function lift<A, B, C, D, E, F, Z>(fn: (a: A, b: B, c: C, d: D, e: E, f: F) => Z, sa: Stream<A>, sb: Stream<B>, sc: Stream<C>, sd: Stream<D>, se: Stream<E>, sf: Stream<F>): Stream<Z>;
+	export function lift<T>(fn: (...values: any[]) => T, ...streams: Array<Stream<any>>): Stream<T>;
+	/** Creates a stream whose value is the array of values from an array of streams. */
+	export function merge(streams: Array<Stream<any>>): Stream<any[]>;
```
[the remark this answers is not in the document]
Comment (spacejack) [reply types/mithril/stream/index.d.ts]: This one isn't working with different stream types.

File: types/mithril/stream/index.d.ts [hunk 7]
```diff
@@ -1,40 +1,49 @@
...
-		/** A co-dependent stream that unregisters dependent streams when set to true. */
-		end: Stream<boolean>;
-		/** When a stream is passed as the argument to JSON.stringify(), the value of the stream is serialized. */
-		toJSON(): string;
-		/** Returns the value of the stream. */
-		valueOf(): T;
-	}
+// tslint:disable:rulename strict-export-declare-modifiers
+/** Creates an empty stream. */
+declare function Stream<T = any>(): Stream<T>; // tslint:disable-line no-unnecessary-generics
```
[the remark this answers is not in the document]
Comment (spacejack) [reply types/mithril/stream/index.d.ts]: We still need the linter disabled here.

File: types/mithril/stream/index.d.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (spacejack) [reply types/mithril/stream/index.d.ts]: The following signature works, however it's wrong:
```ts
export function combine<S extends Stream<any>[], T>(combiner: (...streams: S) => T, streams: {[I in keyof S]: S[I]}): Stream<T>;
```
The final parameter for the combiner (`...streams`) is missing - it should be a `changed` stream. I don't see a way to write that.

File: types/mithril/stream/index.d.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (spacejack) [reply types/mithril/stream/index.d.ts]: Nevermind, solved this one.

**Comment by spacejack** [comment]
I've made all the changes requested except for `combine`. I'm out of ideas on how to make that work. The combiner needs a signature like `(...streams, changed)` but I don't know of a way to write that.

File: types/mithril/stream/index.d.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (isiahmeadows) [reply types/mithril/stream/index.d.ts]: I'm okay with keeping the current type for now, then, until https://github.com/microsoft/TypeScript/issues/1360 gets fixed. I *could* write a valid type that correctly types the parameters here, but the error message if you mismatch one of the `streams` arguments would be *awful*.

**Review by dead-claudia — APPROVED** [review]

**Pull request merged**